### PR TITLE
global: admin views for invenio-accounts

### PR DESCRIPTION
**Pull request by bruno314**
Signed-off-by: Bruno Cuc <bruno.cuc@cern.ch>

### CODE REVIEW

File: examples/app.py [hunk 1]
```diff
@@ -74,6 +75,8 @@
 
 from invenio_accounts import InvenioAccounts
 from invenio_accounts.views import blueprint
+###
```
Comment (jirikuncar) [inline examples/app.py]: ?

File: examples/app.py [hunk 2]
```diff
@@ -105,11 +108,19 @@
 Babel(app)
 Mail(app)
 InvenioDB(app)
+InvenioAccounts(app)
+Menu(app)
```
Comment (jirikuncar) [inline examples/app.py]: you might want to load it before InvenioAccounts

File: invenio_accounts/admin.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        'email',   'last_login_at', 'id'
+    )
+
+    form_overrides = {
+        'last_login_at': DateTimeField
+    }
+
+    column_details_list = column_list
+
+    @action('inactivate', 'Inactivate',
```
Comment (jirikuncar) [inline invenio_accounts/admin.py]: `..., _('Inactivate'),`

File: invenio_accounts/admin.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+            'Are you sure you want to inactivate selected users?')
```
Comment (jirikuncar) [inline invenio_accounts/admin.py]: `_('Are you ...?'))`

File: invenio_accounts/admin.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+
+class AccMixin(object):
+    """Configure access rights."""
+
+    _can_create = True
+    _can_edit = True
+    _can_delete = False
+
+    acc_view_action = 'cfgwebaccess'
```
Comment (lnielsen) [inline invenio_accounts/admin.py]: Needs to be removed.

File: invenio_accounts/admin.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+                    raise ValueError(_("Cannot find user."))
+                if _datastore.activate_user(user_obj):
+                    count += 1
+            if count > 0:
+                flash(_('User(s) were successfully inactivated.', 'success'))
+        except Exception as ex:
+            if not self.handle_view_exception(ex):
+                raise
+
+            flash(_('Failed to activate users. {0}').format(str(ex)),
```
Comment (jirikuncar) [inline invenio_accounts/admin.py]: please check documentation on "Using translations" https://pythonhosted.org/Flask-BabelEx/#using-translations

File: setup.py [hunk 7]
```diff
@@ -144,6 +144,10 @@ def run_tests(self):
     include_package_data=True,
     platforms='any',
     entry_points={
+        'invenio_admin.views': [
+          'invenio_accounts_user = invenio_accounts.admin:user_adminview',
```
Comment (jirikuncar) [inline setup.py]: wrong indentation

File: tests/test_admin.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
+from flask import current_app
```
Comment (jirikuncar) [inline tests/test_admin.py]: missing license

File: invenio_accounts/admin.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
+"""Admin views for invenio-accounts."""
```
Comment (jirikuncar) [inline invenio_accounts/admin.py]: missing license

File: invenio_accounts/admin.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+    form_overrides = {
+        'last_login_at': DateTimeField
+    }
+
+    column_labels = {
+        'current_login_ip': _('Current Login IP'),
+        'last_login_ip': _('Last Login IP')
+    }
+
+    # inline_models = (, )
```
Comment (lnielsen) [inline invenio_accounts/admin.py]: Remove.

File: examples/app.py [hunk 11]
```diff
@@ -22,54 +22,18 @@
 # waive the privileges and immunities granted to it by virtue of its status
 # as an Intergovernmental Organization or submit itself to any jurisdiction.
 
-
-u"""Minimal Flask application example for development.
```
Comment (jirikuncar) [inline examples/app.py]: @bruno314 can you keep and update this little tutorial?

File: setup.py [hunk 12]
```diff
@@ -146,6 +147,10 @@ def run_tests(self):
     include_package_data=True,
     platforms='any',
     entry_points={
+        'invenio_admin.views': [
+            'invenio_accounts_user = invenio_accounts.admin:user_adminview',
+            'invenio_accounts_role = invenio_accounts.admin:role_adminview'
```
Comment (jirikuncar) [inline setup.py]: please add `,` at the end of line

File: invenio_accounts/admin.py [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    @action('inactivate', _('Inactivate'),
+            _('Are you sure you want to inactivate selected users?'))
+    @commit
+    def action_inactivate(self, ids):
+        """Inactivate users."""
+        try:
+            count = 0
+            for user in ids:
```
Comment (jirikuncar) [inline invenio_accounts/admin.py]: IMHO better naming would be:

```python
for user_id in ids:
    user = _datastore.get_user(user)
    ...
```

**Comment by jirikuncar** [comment]
Please check that you have updated copyrights in all files you have touched. Thanks

File: examples/app.py [hunk 14]
```diff
@@ -105,11 +107,19 @@
...
 Mail(app)
 InvenioDB(app)
+Menu(app)
+InvenioAccounts(app)
+
+
+def identity(x):
+    return x
+
+InvenioAdmin(app, permission_factory=identity, view_class_factory=identity)
```
Comment (jirikuncar) [inline examples/app.py]: - `permission_factory=lambda x: x`
- remove identity function

Comment (jirikuncar) [inline examples/app.py]: if INVENIO_ADMIN_AVAILABLE:

File: setup.py [hunk 15]
```diff
@@ -87,6 +87,7 @@
     'Flask-Menu>=0.4.0',
     'Flask-Security>=1.7.5',
     'SQLAlchemy-Utils[ipaddress]>=0.31.0',
+    'invenio-admin>=1.0.0a2'
```
Comment (jirikuncar) [inline setup.py]: it's only extra or tests requirement

File: tests/test_admin.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+    with app.app_context():
+        with app.test_client() as client:
+            inserted_id = _datastore.get_user('test@test.cern').id
+
+            res = client.post(
+                request_url,
+                data={'rowid': inserted_id, 'action': 'activate'},
+                follow_redirects=True
+            )
+
```
Comment (jirikuncar) [inline tests/test_admin.py]: .

File: tests/test_admin.py [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+            assert res.status_code == 200
+
+            res = client.post(
+                request_url,
+                data={'rowid': inserted_id, 'action': 'inactivate'},
+                follow_redirects=True
+            )
+
```
Comment (jirikuncar) [inline tests/test_admin.py]: .

File: tests/test_admin.py [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+            assert res.status_code == 200
+
+            import pytest
+            pytest.raises(ValueError, client.post, request_url,
+                          data={'rowid': -42, 'action': 'inactivate'},
+                          follow_redirects=True
+                          )
```
Comment (jirikuncar) [inline tests/test_admin.py]: .

File: tests/test_admin.py [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+
+            pytest.raises(ValueError, client.post, request_url,
+                          data={'rowid': -42, 'action': 'activate'},
+                          follow_redirects=True
+                          )
```
Comment (jirikuncar) [inline tests/test_admin.py]: .

File: invenio_accounts/admin.py [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+            count = 0
+            for user_id in ids:
+                user = _datastore.get_user(user_id)
+                if user is None:
+                    raise ValueError(_("Cannot find user."))
+                if _datastore.deactivate_user(user):
+                    count += 1
+            if count > 0:
+                flash(_('User(s) were successfully inactivated.'), 'success')
+        except Exception as ex:
```
Comment (jirikuncar) [inline invenio_accounts/admin.py]: `exc`

File: invenio_accounts/admin.py [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+            if not self.handle_view_exception(ex):
+                raise
+
+            flash(_('Failed to inactivate users. %(ex)s', ex=str(ex)),
```
Comment (jirikuncar) [inline invenio_accounts/admin.py]: `exc`

File: invenio_accounts/models.py [hunk 22]
```diff
@@ -87,3 +87,7 @@ def validate_ip(self, key, value):
         if value == 'untrackable':  # pragma: no cover
             value = None
         return value
+
+    def __repr__(self):
```
Comment (jirikuncar) [inline invenio_accounts/models.py]: `__str__` or at least some nicer string info `return 'User<id={0.id}, email={0.email}>'.format(self)`

File: tests/test_admin.py (hunk 19 above)
Comment (bruno314) [inline tests/test_admin.py]: Doesn't pass PEP8 with bracket moved 4 intendation levels left ` E124 closing bracket does not match visual indentation`

File: invenio_accounts/admin.py [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+            count = 0
+            for user_id in ids:
+                user = _datastore.get_user(user_id)
+                if user is None:
+                    raise ValueError(_("Cannot find user."))
+                if _datastore.deactivate_user(user):
+                    count += 1
+            if count > 0:
+                flash(_('User(s) were successfully inactivated.'), 'success')
+        except Exception as exc:
```
Comment (jirikuncar) [inline invenio_accounts/admin.py]: do you really need to catch all exception?

Comment (bruno314) [inline invenio_accounts/admin.py]: Yes.

Comment (bruno314) [inline invenio_accounts/admin.py]: this code is refractored from Zenodo

File: run-tests.sh [hunk 24]
```diff
@@ -25,7 +25,7 @@
 
 pep257 invenio_accounts && \
 check-manifest --ignore ".travis-*" && \
-#isort -rc -c -df **/*.py && \
+# isort -rc -c -df **/*.py && \
```
Comment (jirikuncar) [inline run-tests.sh]: if you are touching this please uncomment it and fix the `.editorconfig [isort]`

**Comment by jirikuncar** [comment]
- [ ] missing access protection /cc @lnielsen

**Pull request merged**